src/core/writer.js

```diff
@@ -281,6 +283,118 @@ function updateXFA({ xfaData, xfaDatasetsRef, newRefs, xref }) {
   newRefs.push({ ref: xfaDatasetsRef, data });
 }
 
+async function getXRefTable(xrefInfo, baseOffset, newRefs, newXref, buffer) {
+  buffer.push("xref\n");
+  let refPosition = -1;
+  let lastRef = null;
+  let refCounter = -1;
+  for (const { ref, data } of newRefs) {
+    if (!lastRef || ref.num !== lastRef.num + refCounter) {
+      if (lastRef) {
+        buffer[refPosition] = `${lastRef.num} ${refCounter}\n`;
+      }
+      lastRef = ref;
+      refPosition = buffer.length;
+      buffer.push("");
+      refCounter = 1;
+    } else {
+      refCounter += 1;
+    }
+    // The EOL is \r\n to make sure that every entry is exactly 20 bytes long.
+    // (see 7.5.4 - Cross-Reference Table).
+    buffer.push(
+      `${baseOffset.toString().padStart(10, "0")} ${Math.min(ref.gen, 0xffff).toString().padStart(5, "0")} n\r\n`
+    );
+    baseOffset += data.length;
+  }
+  if (lastRef) {
+    buffer[refPosition] = `${lastRef.num} ${refCounter}\n`;
+  }
+  computeIDs(baseOffset, xrefInfo, newXref);
+  buffer.push("trailer\n");
+  await writeDict(newXref, buffer);
+  buffer.push("\nstartxref\n", baseOffset.toString(), "\n%%EOF\n");
+}
+
+async function getXRefStreamTable(
+  xrefInfo,
+  baseOffset,
+  newRefs,
+  newXref,
+  buffer
+) {
+  const xrefTableData = [];
+  const indexes = [];
+  let maxOffset = 0;
+  let maxGen = 0;
+  let lastRefNum = NaN;
+  for (const { ref, data } of newRefs) {
+    maxOffset = Math.max(maxOffset, baseOffset);
+    const gen = Math.min(ref.gen, 0xffff);
+    maxGen = Math.max(maxGen, gen);
+    xrefTableData.push([1, baseOffset, gen]);
+    baseOffset += data.length;
+    if (ref.num === lastRefNum + 1) {
+      indexes[indexes.length - 1] += 1;
+      lastRefNum += 1;
+    } else {
+      indexes.push(ref.num, 1);
+      lastRefNum = ref.num;
+    }
+  }
+  newXref.set("Index", indexes);
+  const offsetSize = getSizeInBytes(maxOffset);
+  const maxGenSize = getSizeInBytes(maxGen);
+  const sizes = [1, offsetSize, maxGenSize];
+  newXref.set("W", sizes);
+  computeIDs(baseOffset, xrefInfo, newXref);
+
+  const structSize = sizes.reduce((a, x) => a + x, 0);
+  const data = new Uint8Array(structSize * xrefTableData.length);
+  const stream = new Stream(data);
+  stream.dict = newXref;
+
+  let offset = 0;
+  for (const [type, objOffset, gen] of xrefTableData) {
+    offset = writeInt(type, sizes[0], offset, data);
+    offset = writeInt(objOffset, sizes[1], offset, data);
+    offset = writeInt(gen, sizes[2], offset, data);
+  }
+
+  await writeObject(xrefInfo.newRef, stream, buffer, {});
+  buffer.push("startxref\n", baseOffset.toString(), "\n%%EOF\n");
+}
+
+function computeIDs(baseOffset, xrefInfo, newXref) {
+  if (Array.isArray(xrefInfo.fileIds) && xrefInfo.fileIds.length > 0) {
+    const md5 = computeMD5(baseOffset, xrefInfo);
+    newXref.set("ID", [xrefInfo.fileIds[0], md5]);
+  }
+}
+
+function getTrailerDict(xrefInfo, newRefs, isXrefStream) {
```

Comment:
For overall consistency, can we keep the `useXrefStream` name throughout the code?
```suggestion
function getTrailerDict(xrefInfo, newRefs, useXrefStream) {
```